hail/src/main/scala/is/hail/expr/ir/ExtractIntervalFilters.scala

```diff
@@ -1,360 +1,909 @@
+        case _ => false
       }
-      case _ => isRowRef(struct) && fieldName == keyFieldName
     }
 
-    def isFirstKey(ir: IR): Boolean = isKeyField(ir, rowKeyType.fieldNames.head)
+    object BoolValue {
+      private def all: KeySet = KeySetLattice.top
+      private def none: KeySet = KeySetLattice.bottom
 
-    def isKeyStructPrefix(ir: IR): Boolean = ir match {
-      case MakeStruct(fields) =>
-        fields.size <= rowKeyType.size && fields.view.zipWithIndex.forall { case ((_, fd), idx) =>
-          isKeyField(fd, rowKeyType.fieldNames(idx))
+      def allTrue(keySet: KeySet = KeySetLattice.top): BoolValue = ConcreteBool(keySet, none, none)
+      def allFalse(keySet: KeySet = KeySetLattice.top): BoolValue = ConcreteBool(none, keySet, none)
+      def allNA(keySet: KeySet = KeySetLattice.top): BoolValue = ConcreteBool(none, none, keySet)
+      def top(keySet: KeySet = KeySetLattice.top): BoolValue = ConcreteBool(keySet, keySet, keySet)
+
+      def apply(trueBound: KeySet, falseBound: KeySet, naBound: KeySet): BoolValue = {
+        if (trueBound == all && falseBound == all && naBound == all)
+          Lattice.top
+        else if (trueBound == none && falseBound == none && naBound == none)
+          Lattice.bottom
+        else
+          ConcreteBool(trueBound, falseBound, naBound)
+      }
+
+      def or(l: BoolValue, r: BoolValue): BoolValue = (l, r) match {
+        case (ConstantBool(l, lKeySet), ConstantBool(r, rKeySet)) =>
+          assert(lKeySet eq rKeySet)
+          ConstantBool(l || r, lKeySet)
+        case _ => ConcreteBool(
+          KeySetLattice.combine(l.trueBound, r.trueBound),
+          KeySetLattice.meet(l.falseBound, r.falseBound),
+          KeySetLattice.combineMulti(
+            KeySetLattice.meet(l.naBound, r.falseBound),
+            KeySetLattice.meet(l.naBound, r.naBound),
+            KeySetLattice.meet(l.falseBound, r.naBound)))
+      }
+
+      def and(l: BoolValue, r: BoolValue): BoolValue = (l, r) match {
+        case (ConstantBool(l, lKeySet), ConstantBool(r, rKeySet)) =>
+          assert(lKeySet eq rKeySet)
+          ConstantBool(l && r, lKeySet)
+        case _ => ConcreteBool(
+          KeySetLattice.meet(l.trueBound, r.trueBound),
+          KeySetLattice.combine(l.falseBound, r.falseBound),
+          KeySetLattice.combineMulti(
+            KeySetLattice.meet(l.naBound, r.trueBound),
+            KeySetLattice.meet(l.naBound, r.naBound),
+            KeySetLattice.meet(l.trueBound, r.naBound)))
+      }
+
+      def not(x: BoolValue): BoolValue = x match {
+        case ConstantBool(x, keySet) => ConstantBool(!x, keySet)
+        case _ => ConcreteBool(x.falseBound, x.trueBound, x.naBound)
+      }
+
+      def isNA(x: BoolValue): BoolValue = x match {
+        case ConstantBool(x, keySet) => ConstantBool(false, keySet)
+        case _ => ConcreteBool(x.naBound, KeySetLattice.combine(x.trueBound, x.falseBound), KeySetLattice.bottom)
```

Comment:
This line is duplicated in `BoolValue.isNA()`

Reply:
Whoops, started moving this to a method but didn't finish.